openrr-remote/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+where
+    C: arci::MoveBase + 'static,
+{
+    async fn send_velocity(
+        &self,
+        request: tonic::Request<pb::BaseVelocity>,
+    ) -> Result<tonic::Response<()>, tonic::Status> {
+        let request = request.into_inner();
+        arci::MoveBase::send_velocity(&self.inner, &request.into())
+            .map_err(|e| tonic::Status::unknown(e.to_string()))?;
+        Ok(tonic::Response::new(()))
+    }
+
+    async fn current_velocity(
+        &self,
+        _: tonic::Request<()>,
+    ) -> Result<tonic::Response<pb::BaseVelocity>, tonic::Status> {
+        let v = arci::MoveBase::current_velocity(&self.inner)
+            .map_err(|e| tonic::Status::unknown(e.to_string()))?
+            .into();
+        Ok(tonic::Response::new(v))
+    }
+}
+
+// =============================================================================
+// arci::Navigation
+
+remote_types!(Navigation {
+    client: RemoteNavigationSender,
+    server: RemoteNavigationReceiver,
+});
+
+impl arci::Navigation for RemoteNavigationSender {
+    fn send_goal_pose(
+        &self,
+        goal: arci::Isometry2<f64>,
+        frame_id: &str,
+        timeout: std::time::Duration,
+    ) -> Result<arci::WaitFuture, arci::Error> {
+        let mut client = self.client.clone();
+        let frame_id = frame_id.to_owned();
+        Ok(wait_from_handle(tokio::spawn(async move {
+            client
+                .send_goal_pose(pb::GoalPoseRequest {
+                    goal: Some(goal.into()),
+                    frame_id,
+                    timeout: Some(timeout.into()),
+                })
+                .await
+        })))
+    }
+
+    fn cancel(&self) -> Result<(), arci::Error> {
+        let mut client = self.client.clone();
+        block_in_place(client.cancel(())).map_err(|e| arci::Error::Other(e.into()))?;
+        Ok(())
+    }
+}
+
+#[tonic::async_trait]
+impl<C> navigation_server::Navigation for RemoteNavigationReceiver<C>
+where
+    C: arci::Navigation + 'static,
+{
+    async fn send_goal_pose(
+        &self,
+        request: tonic::Request<pb::GoalPoseRequest>,
+    ) -> Result<tonic::Response<()>, tonic::Status> {
+        let request = request.into_inner();
+        arci::Navigation::send_goal_pose(
+            &self.inner,
+            request.goal.unwrap().into(),
+            &request.frame_id,
+            request.timeout.unwrap().try_into().unwrap(),
+        )
+        .map_err(|e| tonic::Status::unknown(e.to_string()))?
+        .await
+        .map_err(|e| tonic::Status::unknown(e.to_string()))?;
+        Ok(tonic::Response::new(()))
+    }
+
+    async fn cancel(&self, _: tonic::Request<()>) -> Result<tonic::Response<()>, tonic::Status> {
+        arci::Navigation::cancel(&self.inner).map_err(|e| tonic::Status::unknown(e.to_string()))?;
+        Ok(tonic::Response::new(()))
+    }
+}
+
+// =============================================================================
```

Comment:
I think it's better to divide into multiple files if you need this ================

Reply:
(We won't need to stick to the file layout issue here since we will eventually do something like https://github.com/openrr/openrr/pull/431.)